src/lib.rs

```diff
@@ -376,26 +359,36 @@ where
             });
         };
         let tt = task_token.clone();
-        match status {
-            activity_result::Status::Completed(ar::Success { result }) => {
-                self.server_gateway
-                    .complete_activity_task(task_token, result.map(Into::into))
-                    .await?;
-            }
-            activity_result::Status::Failed(ar::Failure { failure }) => {
-                self.server_gateway
-                    .fail_activity_task(task_token, failure.map(Into::into))
-                    .await?;
-            }
-            activity_result::Status::Canceled(ar::Cancelation { details }) => {
-                self.server_gateway
-                    .cancel_activity_task(task_token, details.map(Into::into))
-                    .await?;
+        let maybe_net_err = match status {
+            activity_result::Status::Completed(ar::Success { result }) => self
+                .server_gateway
+                .complete_activity_task(task_token, result.map(Into::into))
+                .await
+                .err(),
+            activity_result::Status::Failed(ar::Failure { failure }) => self
+                .server_gateway
+                .fail_activity_task(task_token, failure.map(Into::into))
+                .await
+                .err(),
+            activity_result::Status::Canceled(ar::Cancelation { details }) => self
+                .server_gateway
+                .cancel_activity_task(task_token, details.map(Into::into))
+                .await
+                .err(),
+        };
+        let (res, should_remove) = match maybe_net_err {
+            Some(e) if e.code() == tonic::Code::NotFound => {
+                warn!(task_token = ?tt, details = ?e, "Activity not found on completion");
+                (Ok(()), true)
             }
+            Some(err) => (Err(err), false),
+            None => (Ok(()), true),
+        };
+        if should_remove {
+            self.outstanding_activity_tasks.remove(&tt);
```

Comment:
What is the implication of removing the activity task?
What happens if the activity tries to send heartbeats after this removal?